RES-1652 Sentry

edwh opened the pull request:
There's a bug on the page for a past event, where you can click the "Add volunteer" button, and then click to submit without selecting a user.  This creates an error on the server.

I've converted this modal to Vue, and introduced a couple of API calls to handle retrieving the list of volunteers from a group, and adding a volunteer.  

These are the first new API calls which require authentication.  

The cheap and cheerful Laravel way of doing API authentication is via the URL `api_token` parameter.
Not all users necessarily have an API token set up.  But it's time they did.  There is an existing `ensureAPIToken` method, so we add an `EnsureAPIToken `middleware to invoke this on logged-in users, and change the factory so that test users always do.

We then need a way for the Vue client to pass this token to the server.  Return this in an unencrypted cookie from the server, and update our `auth` mixin to grab that and make it available.  This replaces the use of props for API tokens.

The upshot of this is that the API is now available for logged in users.  In the long term we should be looking at Laravel Passport and JWTs, but I don't think it makes sense to do that yet.

Other stuff:
* The API controller is called EventController whereas the web one is called PartyController.  Rename it to match.
* Delete `checkAPIAccess `middleware which was only used by API calls which have now been retired.
* Add tests for CalendarEventsController and fix a couple of very minor bugs.

ngm commented:
@edwh 

> The upshot of this is that the API is now available for logged in users. 

Nice!

> The cheap and cheerful Laravel way of doing API authentication is via the URL api_token parameter.

In case it's of use, for reference the cheap and cheerful way also allows the token to come via form params or bearer token.  https://laravel.com/docs/6.x/api-authentication

> Not all users necessarily have an API token set up. But it's time they did. 

We need to consider API permissions and scopes more carefully now, then.  (This is a good thing!)  But previously access control to the API was controlled blanketly by the presence of a token.  No token, no access.  With the change here anyone could access the API if they can figure out their token.  This sounds fairly trivial to do, if it's in an unencrypted cookie.  

So, if they can access the API, what can they do?  Nothing destructive at present.  And as far as data they could access, I believe this is relatively well permissioned already - e.g. if you try to access info on groups and events for a network, you only get access to your group.  There are a couple of sensitive API endpoints we should double-check are well permissioned already, e.g. anything accessing user information.

Additionally, what's the possibility of someone figuring out someone else's token?  I imagine it's possible somehow with some kind of man-in-the-middle, but pretty low risk?  Basically the situation we need to avoid is the ability to access a token that then allows access to sensitive data.

>  This replaces the use of props for API tokens.

Great!

> In the long term we should be looking at Laravel Passport and JWTs, but I don't think it makes sense to do that yet.

I think we should do a spike on this next sprint, as I'm interested how it might work and if provides any useful affordances for scopes/permissions.

Will take a look at the code now

edwh commented:
> > The cheap and cheerful Laravel way of doing API authentication is via the URL api_token parameter.
> 
> In case it's of use, for reference the cheap and cheerful way also allows the token to come via form params or bearer token. https://laravel.com/docs/6.x/api-authentication

Yes.  The tricky part is getting the token to the API user, really.

> 
> > Not all users necessarily have an API token set up. But it's time they did.
> 
> We need to consider API permissions and scopes more carefully now, then.

We do.  Added some logic to UserController::getUserList.  I can't see any other issues but please look.

> 
> Additionally, what's the possibility of someone figuring out someone else's token? I imagine it's possible somehow with some kind of man-in-the-middle, but pretty low risk? Basically the situation we need to avoid is the ability to access a token that then allows access to sensitive data.

I don't think the risk of accessing sensitive data is much increased by this, because the API users who could access sensitive data had tokens.  The token is not fantastically secure - just a random 60 character string.  But I think that's ok for now.

> > In the long term we should be looking at Laravel Passport and JWTs, but I don't think it makes sense to do that yet.
> 
> I think we should do a spike on this next sprint, as I'm interested how it might work and if provides any useful affordances for scopes/permissions.

Agree.

edwh commented on app/Http/Controllers/API/GroupController.php:
```diff
@@ -219,4 +221,31 @@ public static function getEventsForGroup(Request $request, Group $group)
+        $user = auth('api')->user();
+
+        if (!$user || !Fixometer::userHasEditGroupPermission($idgroups, $user->id)) {
+            // We require host permissions to view the list of volunteers.
```
Yes, the API call is only used in that case, from within the popup modal.  There's an argument that you should be able to see some details - but not all, e.g. email - with other permissions.  But I think we can defer that until we use the call more widely.  Will comment.

edwh commented on routes/web.php:
```diff
@@ -341,64 +341,66 @@
-    if (Fixometer::hasRole(Auth::user(), 'Administrator')) {
-        return view('mediawiki.index');
-    }
+    Route::get('/media-wiki', function () {
```
It does - the view doesn't seem to exist any more.  Deleted.

edwh commented:
@ngm I've added some tests for the fixes made recently to this branch.  Now that we're using expandEvent more widely, the client can use event.group rather than event.the_group.

I've also fixed some test issues exposed by DST change.

I think it's worth you giving this another look/test before we merge, since it's now quite a bit change.  But if you're short of time we can just merge it into develop and handle any issues later.

edwh commented on routes/web.php:
```diff
@@ -11,168 +11,169 @@
-    Route::get('/thumbnail/', 'UserController@getThumbnail');
-    Route::get('/menus/', 'UserController@getUserMenus');
-});
+Route::group(['middleware' => ['ensureAPIToken']], function () {
```
Exactly.
